### PR TITLE
Feature/ethereum chain interface

### PR DESCRIPTION
**Pull request by basarrcan**

### CODE REVIEW

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  isValidEthereumPrivateKey,
+} from './cryptoModelHelpers'
+
+export function isValidEthereumDateString(str: string): str is EthereumDate {
+  if (isNullOrEmpty(str)) return false
+  return str.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{2}\d+$/i) !== null
+}
+
+export function isValidEthereumAsset(symbol: string): boolean {
+  if (
```
Comment (surabhil) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:19]: I guess you can create an array of all symbols and then code it like if(ethereumAssetSymbols.include(symbol)

or use EThUnit from cryptoModels

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    symbol === 'mether' ||
+    symbol === 'gether' ||
+    symbol === 'tether'
+  )
+    return true
+  return false
+}
+
+export function isValidEthereumEntityName(str: EthereumEntityName | string): str is EthereumEntityName {
+  return !!(
```
Comment (traylewin) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:53]: !! not needed

File: src/chains/ethereum_1/ChainEthereumV1.ts [hunk 3]
```diff
@@ -135,17 +151,53 @@ class ChainEthereumV1 implements Chain {
 
   // Chain Helper Functions
 
-  isValidEntityName = notImplemented
+  // --------- Chain helper functions
+
+  /** Verifies that the value is a valid chain entity name (e.g. an account name) */
+  public isValidEntityName = (value: string): boolean => {
+    return !!isValidEthereumEntityName(value)
```
Comment (traylewin) [inline src/chains/ethereum_1/ChainEthereumV1.ts:158]: The !! operator transforms any value into boolean. If the source value is already a boolean, then you don't need to do !!. So this function should just return isValidEthereumEntityName(value) not !!isValidEthereumEntityName(value)

File: src/chains/ethereum_1/ChainEthereumV1.ts [hunk 4]
```diff
@@ -135,17 +151,53 @@ class ChainEthereumV1 implements Chain {
...
+  }
+
+  /** Verifies that the value is a valid EOS entity name (e.g. an account name) */
+  isValidEthereumEntityName = isValidEthereumEntityName
+
+  /** Verifies that the value is a valid chain asset string */
+  public isValidAsset = (value: string): boolean => {
+    return !!isValidEthereumAsset(value)
```
Comment (traylewin) [inline src/chains/ethereum_1/ChainEthereumV1.ts:166]: !! not needed

File: src/chains/ethereum_1/ChainEthereumV1.ts [hunk 5]
```diff
@@ -135,17 +151,53 @@ class ChainEthereumV1 implements Chain {
...
+  }
+
+  /** Verifies that the value is a valid EOS asset string */
+  isValidEthereumAsset = isValidEthereumAsset
 
-  isValidAsset = notImplemented
+  /** Verifies that the value is a valid chain date */
+  public isValidDate = (value: string): boolean => {
+    return !!isValidEthereumDateString(value)
```
Comment (traylewin) [inline src/chains/ethereum_1/ChainEthereumV1.ts:174]: !! not needed

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  if (isValidEthereumAsset(symbol)) throw new Error('Symbol must be ethereum unit type')
+  return toWei(amount, toEthUnit(symbol)) as EthereumAsset
+}
+
+export function toEthereumEntityName(name: string): EthereumEntityName {
+  if (isValidEthereumEntityName(name)) {
+    return name
+  }
+
+  if (name === '') {
```
Comment (surabhil) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:86]: May be use isNullOrEmpty here

Comment (traylewin) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:86]: this should use isNullOrEmpty to cover 'undefined' and other edge cases

File: src/chains/ethereum_1/ChainEthereumV1.ts [hunk 7]
```diff
@@ -135,17 +151,53 @@ class ChainEthereumV1 implements Chain {
...
+
+  /** Verifies that the value is a valid EOS entity name (e.g. an account name) */
+  isValidEthereumEntityName = isValidEthereumEntityName
+
+  /** Verifies that the value is a valid chain asset string */
+  public isValidAsset = (value: string): boolean => {
+    return !!isValidEthereumAsset(value)
+  }
+
+  /** Verifies that the value is a valid EOS asset string */
```
Comment (traylewin) [inline src/chains/ethereum_1/ChainEthereumV1.ts:169]: Comment is wrong - update (or add) comments to all functions

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  } else {
+    const dateString = moment(date).format('YYYY-MM-DDTHH:MM:SS.sss')
+    if (isValidEthereumDateString(dateString)) {
+      return dateString
+    }
+  }
+  throw new Error(`Should not get here. (invalid toDateStr provided): ${date}`)
+}
+
+export function toEthereumAsset(amount: number, symbol: string): EthereumAsset {
```
Comment (traylewin) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:76]: Add comment to describe expected ins and outs

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  isValidEthereumTxData,
+  isValidEthereumPrivateKey,
+} from './cryptoModelHelpers'
+
+export function isValidEthereumDateString(str: string): str is EthereumDate {
+  if (isNullOrEmpty(str)) return false
+  return str.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{2}\d+$/i) !== null
+}
+
+export function isValidEthereumAsset(symbol: string): boolean {
```
Comment (traylewin) [inline src/chains/ethereum_1/helpers/generalModelHelpers.ts:18]: This function is confusingly named and probably not what is expected. isValidEthereumAsset should be used to test an asset param assigned for an ETH contract/action.
This function as written does something different - it checks whether a string is a valid ETH asset unit type. I would change the param from 'symbol' to 'unitName' and rename this function isValidEthereumAssetUnit.
Then I would add a new function isValidEthereumAsset that checks a value to see if it is a well-formed asset value - which is just a positive number. Right?

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts (hunk 1 above)
Comment (traylewin) [reply src/chains/ethereum_1/helpers/generalModelHelpers.ts:19]: ah yes. The units should not be duplicated here - they are in EthUnit. This function should still accept a string but then check that the value against valid options in EthUnit
Also ethUnit should be an enum not a type. Then you can copy the isInEnum function from oreid-service into chainjs helpers and use it to check if a value is in the enum

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts (hunk 9 above)
Comment (basarrcan) [reply src/chains/ethereum_1/helpers/generalModelHelpers.ts]: Must be clear now

File: src/chains/ethereum_1/helpers/generalModelHelpers.ts (hunk 2 above)
Comment (traylewin) [reply src/chains/ethereum_1/helpers/generalModelHelpers.ts:30]: Is this !! still there? It's not needed

File: src/chains/ethereum_1/ethErrors.ts [hunk 10]
```diff
@@ -11,7 +11,7 @@ export const ChainErrorRegExs: { [key: string]: string } = {
 }
 
 // Maps an Error object (thrown by a call to the chain) into a known set of errors
-// RpcError is an eosjs structure that includes the 'json' property that has error details
+// RpcError is an web3js structure that includes the 'json' property that has error details
```
Comment (traylewin) [inline src/chains/ethereum_1/ethErrors.ts:14]: This comment is not correct. The Rpcerror object is something from eosjs and isn't in web3.
Also, we should visit how errors are working from web3 - I suspect we need to change something here

**Pull request merged**